cocaine/detail/secadaptor.py

```diff
@@ -0,0 +1,152 @@
+'''Secure service adaptor.
+
+Provides method for Service adaptor construction, capable to wrap Cocaine
+service and inject security token into the header of service request.
+
+Note: mostly copy-pasted from cocaine-tools `SecureService` implementation.
+'''
+import time
+
+from tornado import gen
+
+from .service import Service
+from ..exceptions import CocaineError
+
+
+class SecureServiceError(CocaineError):
+    pass
+
+
+class Promiscuous(object):
+    '''Null token fetch interface implementation.
+
+    Used for fallback in case of unsupported (not set) secure module type
+    provided by user, access errors due empty token will be propagated
+    to caller code.
+    '''
+    @gen.coroutine
+    def fetch_token(self):
+        raise gen.Return('')
+
+
+class TVM(object):
+    '''Tokens fetch interface implementation.
+
+    Provides public `fetch_token` method, which should be common
+    among various token backends types.
+
+    Attributes:
+        TYPE (str): String representation of token fetcher type,
+            will be included in secure header.
+    '''
+    # Can be taken from class name in case of TVM, but could be inconvenient
+    # in less general name formatting rules.
+    TYPE = 'TVM'
+
+    def __init__(self, client_id, client_secret, name='tvm'):
+        '''TVM
+
+        Args:
+            client_id (int): Integer client identificator.
+            client_secret (str): Client secret.
+            name (str): TVM service name, defaults to 'tvm'.
+        '''
+        self._client_id = client_id
+        self._client_secret = client_secret
+
+        self._tvm = Service(name)
+
+    @gen.coroutine
+    def fetch_token(self):
+        '''Gains token from secure backend service.
+        Returns:
+            token (str): formatted for cocaine protocol header token.
```

Comment:
AFAIK in google docstrings `Returns` section usually write in follow format `type: description`